lib/client/conntest/kube.go

```diff
@@ -0,0 +1,279 @@
+	ProxyHostPort string
+
+	// KubernetesPublicProxyAddr is the kubernetes proxy address.
+	KubernetesPublicProxyAddr string
+
+	// TLSRoutingEnabled indicates that proxy supports ALPN SNI server where
+	// all proxy services are exposed on a single TLS listener (Proxy Web Listener).
+	TLSRoutingEnabled bool
+}
+
+// KubeConnectionTester implements the ConnectionTester interface for Testing Kubernetes access.
+type KubeConnectionTester struct {
+	cfg          KubeConnectionTesterConfig
+	webProxyAddr string
+}
+
+// NewKubeConnectionTester returns a new KubeConnectionTester
+func NewKubeConnectionTester(cfg KubeConnectionTesterConfig) (*KubeConnectionTester, error) {
+	parsedProxyHostAddr, err := client.ParseProxyHost(cfg.ProxyHostPort)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &KubeConnectionTester{
+		cfg:          cfg,
+		webProxyAddr: parsedProxyHostAddr.WebProxyAddr,
+	}, nil
+}
+
+// TestConnection tests an Kubernetes Access to the target Kubernetes Cluster using
+//   - the provided client
+//   - resource name
+//
+// A new ConnectionDiagnostic is created and used to store the traces as it goes through the checkpoints
+// To connect to the KubeCluster, we will create a cert-key pair and setup a Kubernetes client back to Teleport Proxy.
+// We report the following cases:
+//   - trace of whether the Kubernetes cluster is reachable
+//   - trace of whether the User Role defines Kubernetes principals for the cluster: `kubernetes_groups` & `kubernetes_users`
+//   - trace of whether the User role has access to the desired kubernetes cluster: `kubernetes_labels` allow access.
+//   - trace of weather the cluster is accessible and we can list pods on the desired namespace.
+func (s *KubeConnectionTester) TestConnection(ctx context.Context, req TestConnectionRequest) (types.ConnectionDiagnostic, error) {
+	if req.ResourceKind != types.KindKubernetesCluster {
+		return nil, trace.BadParameter("invalid value for ResourceKind, expected %q got %q", types.KindKubernetesCluster, req.ResourceKind)
+	}
+
+	connectionDiagnosticID := uuid.NewString()
+	connectionDiagnostic, err := types.NewConnectionDiagnosticV1(connectionDiagnosticID, map[string]string{},
+		types.ConnectionDiagnosticSpecV1{
+			// We start with a failed state so that we don't need to set it to each return statement once an error is returned.
+			// if the test reaches the end, we force the test to be a success by calling
+			// 	connectionDiagnostic.SetMessage(types.DiagnosticMessageSuccess)
+			//	connectionDiagnostic.SetSuccess(true)
+			Message: types.DiagnosticMessageFailed,
+		})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if err := s.cfg.UserClient.CreateConnectionDiagnostic(ctx, connectionDiagnostic); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	currentUser, err := s.cfg.UserClient.GetCurrentUser(ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	key, err := client.GenerateRSAKey()
```

Comment:
Can this logic be cleaned up a bit, for example I would consider factoring out parts of this function into dedicated method e.g . "createKubeClient", etc.